### ISSUE TITLE
Circular colorbar for azimuthal data

### ISSUE THREAD

**Issue by YaraRossi** [GenericMappingTools/pygmt#3638]
### Description of the desired feature

For various datasets a linear colorbar is appropriate. However when azimuthal data is plotted [0 - 360°] is would make sense to have a circular colorbar that actually represents the direction the data is plotted. This could either be set on the side of the plot or as an inset. At the moment I don't think this exists yet and my only work around would be to save a *.png figure and then add it to the pygmt plot through fig.image() function. 

An example is the direction of tilt. It's possible to visualize this with vectors, but when the directions get more heterogeneous it gets messy. This is an example of where the direction of the ground tilt is visualized according to azimuth. The linear colorbar is confusing, as the colors are actually continuous and azimuthal.

<img width="502" alt="Screenshot 2024-11-20 at 14 42 17" src="https://github.com/user-attachments/assets/830a47c0-bcfe-4427-ae7d-32ba37af7f52">

the colorbar (colorcircle?) should be something like this:

<img width="367" alt="Screenshot 2024-11-20 at 14 43 26" src="https://github.com/user-attachments/assets/3d827aee-6dde-4724-b05c-01e48597525a">

### Are you willing to help implement and maintain this feature?

Yes

**Comment by weiji14** [comment]
I'm not sure if there's a way to get a circular colorbar, but GMT/PyGMT does have cyclic colormaps listed at https://docs.generic-mapping-tools.org/6.5/reference/cpts.html#id4, and if you use those, there will be a circle/cycle symbol that denotes it is cyclic:

```python
import pygmt

fig = pygmt.Figure()
pygmt.makecpt(cmap="SCM/corkO", series=[0, 360, 45], continuous=True)
fig.colorbar(cmap=True, region=[0,2,4,8], projection="X10c")
fig.show()
```
produces
![cyclic_colorbar](https://github.com/user-attachments/assets/502258d3-bd57-4c99-8f41-c84be12f409c)

If you want an actual circular colorbar though, that would need to be done in upstream GMT repo https://github.com/GenericMappingTools/gmt, or

**Comment by YaraRossi** [comment]
Thanks for the input. However, I think to understand the data better and faster, and actual circle would be really helpful. I now made a workaround that creates a *.png file with the circular colorbar. That figure can afterwards be inserted into the pygmt figure through fig.image(). Attached is an example file and the figure with the real data. In case anyone else needs the same feature.

Example file:
[Circular_colormap_example.txt](https://github.com/user-attachments/files/17838965/Circular_colormap_example.txt)

Figure:
<img width="427" alt="Screenshot 2024-11-20 at 18 36 01" src="https://github.com/user-attachments/assets/b99aeecb-53ed-47fd-a1c4-c47b045c01a4">

**Comment by weiji14** [comment]
Yes, I agree that a circular colorbar would be more appropriate for your case.

That said, I would advise you to use a proper cyclic colormap, because you current one has a discontinuity at 0 degree / North (the color jumps from green to blue abruptly). Since you are already using a Scientific Colour Map (tofino), you might want to consider one of the cyclic colormaps at https://www.fabiocrameri.ch/colourmaps/ where the transition from 359degrees to 1 degree is smoother. Specifically `romaO`, `bamO`, `brocO`, `corkO`, or `vikO`

![image](https://github.com/user-attachments/assets/c73bfdae-28c7-48e6-b3a6-60259d5c480b)

**Comment by seisman** [comment]
In PyGMT, the circular colorbar can be done by plotting wedges via `Figure.plot` with `style="W"`.

**Comment by yvonnefroehlich** [comment]
Some time ago I created a circular colorbar in PyGMT:
https://github.com/yvonnefroehlich/gmt-pygmt-plotting/pull/23/files

**Comment by YaraRossi** [comment]
Thank you [weiji14](https://github.com/weiji14) you are right, that looks much better!

**Comment by michaelgrund** [comment]
As @seisman mentioned it's also easy to do that by plotting wedges within a 360° range. Also adding a hole in the center without needing any transparency is straightforward.

I experimented with wedges the last few weeks and maybe it's worth to consider if we want to implement some high level functions like circular colorbars, pie charts, doughnut plots etc. based on using wedges with `plot`. I hope I have time to expand my POCs during the holiday season.

Related to #2797.  

```
import pygmt
import numpy as np

def colorbar_az(cmap = "romaO", center_x = 1, center_y = 1, radius_inner = 1.5, radius_outer = 2.5):
    
    values = np.full(360, 1)
    # calc fraction within full circle
    fracs = [value / 360 * 100 for value in values]
    
    # scale colormap
    pygmt.makecpt(cmap=cmap, series=[0, len(values)-1, 1])
    
    start_angle = 0

    for i, frac in enumerate(fracs):
        end_angle = start_angle + (frac / 100) * 360
        fig.plot(x = center_x, 
                 y = center_y, 
                 style = f"w{radius_outer}/{start_angle}/{end_angle}+i{radius_inner}", 
                 zvalue = i, 
                 fill = "+z", 
                 cmap = True
                )
        start_angle = end_angle


fig = pygmt.Figure()

fig.coast(
    region=[-125, -122, 47, 49],
    projection="M6c",
    land="grey",
    water="lightblue",
    shorelines=True,
    frame="a",
)

colorbar_az(center_x = -123, center_y = 48)

fig.show()
```

![plot_cbar_az](https://github.com/user-attachments/assets/4c7111a2-4637-407d-b49e-bd783a73d651)

**Comment by seisman** [comment]
This is another simplified version based on @yvonnefroehlich's script:
```python
import numpy as np
import pygmt

def colorbar_az(cmap="romaO", x0=0, y0=0, diameter_inner=1.5, diameter_outer=2.5, step=1.0):
    pygmt.makecpt(cmap=cmap, series=[0, 360])

    angles = np.arange(0, 360, step)
    # Create the data for the wedges
    data_wedges = np.column_stack([
        np.full_like(angles, x0),  # x
        np.full_like(angles, y0),  # y
        angles,  # For color
        angles,  # start_angle
        angles + step  # end_angle
    ])

    # Plot the all wedges with one call of Figure.plot()
    fig.plot(data=data_wedges, style=f"w{diameter_outer}c+i{diameter_inner}c", cmap=True)

fig = pygmt.Figure()
fig.basemap(region=[-5, 5, -5, 5], frame=True, projection="X10c/10c")
colorbar_az()
fig.show()
```

**Comment by YaraRossi** [comment]
I added another variable that allows to rotate the color, so that 0 can be defined where ever people need it. Additionally, it would be great to have lables for example E, N, W, S, but I couldn't manage to locate them next to the wedges automatically. With fig.text() they stayed in the cartesian coordinate system.

`import numpy as np
import pygmt

def colorbar_az( fig='fig', cmap="romaO", x0=0, y0=0, diameter_inner=1.5, diameter_outer=2.5, step=1.0, start_azimuth = '0:East, 90:North, 180:West, 270:South'):
    import pygmt

    pygmt.makecpt(cmap=cmap, series=[0, 360])

    angles = np.arange(0, 360, step)
    # Create the data for the wedges
    data_wedges = np.column_stack([
        np.full_like(angles, x0),  # x
        np.full_like(angles, y0),  # y
        angles,  # For color
        angles + start_azimuth,  # start_angle
        angles + step + start_azimuth  # end_angle
    ])

    # Plot the all wedges with one call of Figure.plot()
    fig.plot(data=data_wedges, style=f"w{diameter_outer}c+i{diameter_inner}c", cmap=True)`
